ddraw/IDirect3DDeviceX.cpp

```diff
@@ -2453,11 +2542,98 @@ HRESULT m_IDirect3DDeviceX::DrawIndexedPrimitive(D3DPRIMITIVETYPE dptPrimitiveTy
 		}
 		else
 		{
+			const UINT stride = GetVertexStride(dwVertexTypeDesc);
+
+			// Handle PositionT
+			if((dwVertexTypeDesc & D3DFVF_XYZRHW) != 0 && Config.DdrawConvertHomogeneousW)
+			{
+				if(!Config.DdrawConvertHomogeneousToWorld)
+				{
+					UINT8 *vertex = (UINT8*)lpVertices;
+
+					for (UINT x = 0; x < dwVertexCount; x++)
+					{
+						float *pos = (float*) vertex;
+
+						pos[3] = 1.0f;
+
+						vertex += stride;
+					}
+
+					// Update the FVF
+					dwVertexTypeDesc = (dwVertexTypeDesc & ~D3DFVF_XYZRHW) | D3DFVF_XYZW;
+				}
+				else
+				{
+					const UINT newstride = stride - sizeof(float);
+					const UINT restSize = stride - sizeof(float) * 4;
+
+					RenderData.DdrawConvertHomogeneousToWorld_IntermediateGeometry.resize(newstride * dwVertexCount);
+
+					UINT8 *sourceVertex = (UINT8*)lpVertices;
+					UINT8 *targetVertex = (UINT8*)RenderData.DdrawConvertHomogeneousToWorld_IntermediateGeometry.data();
+
+					lpVertices = targetVertex;
+
+					for (UINT x = 0; x < dwVertexCount; x++)
+					{
+						// Transform the vertices into world space
+						float *srcpos = (float*) sourceVertex;
+						float *trgtpos = (float*) targetVertex;
+
+						DirectX::XMVECTOR xpos = DirectX::XMVectorSet(srcpos[0], srcpos[1], srcpos[2], srcpos[3]);
+
+						DirectX::XMVECTOR xpos_global = DirectX::XMVector3TransformCoord(xpos, RenderData.DdrawConvertHomogeneousToWorld_ViewMatrixInverse);
+
+						xpos_global = DirectX::XMVectorDivide(xpos_global, DirectX::XMVectorSplatW(xpos_global));
+
+						trgtpos[0] = DirectX::XMVectorGetX(xpos_global);
+						trgtpos[1] = DirectX::XMVectorGetY(xpos_global);
+						trgtpos[2] = DirectX::XMVectorGetZ(xpos_global);
+
+						// Copy the rest
+						std::memcpy(targetVertex + sizeof(float) * 3, sourceVertex + sizeof(float) * 4, restSize);
+
+						// Move to next vertex
+						sourceVertex += stride;
+						targetVertex += newstride;
+					}
+
+					// Set transform
+					(*d3d9Device)->SetTransform(D3DTS_VIEW, &RenderData.DdrawConvertHomogeneousToWorld_ViewMatrix);
+					(*d3d9Device)->SetTransform(D3DTS_PROJECTION, &RenderData.DdrawConvertHomogeneousToWorld_ProjectionMatrix);
+
+					// Update the FVF
+					const DWORD newVertexTypeDesc = (dwVertexTypeDesc & ~D3DFVF_XYZRHW) | D3DFVF_XYZ;
+
+					// Set fixed function vertex type
+					(*d3d9Device)->SetFVF(newVertexTypeDesc);
```

Comment:
Since here SetFVF overwrites the value that the client last set, other draw calls after this one might still expect the original value to still be active. To support those use-cases, best remember the currently active value before overwriting it here, and restore the value after the draw call below.

Reply:
Unlike Direct3D9, in Direct3D7 and older every draw call has the FVF as an argument.  So `SetFVF()` will be called before every draw call.  So there is never a need to restore the value.

Note: this is even true of Direct3D8.  It is only starting with Direct3D9 that you would need to restore the value of FVF.